modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,382 @@
+  // Add elements to the new mesh
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+  for (unsigned int i = 0; i < _peripheral_layer_num; ++i)
+    for (unsigned int j = 0; j < input_ext_node_num; ++j)
+    {
+      Elem * elem_Quad4 = mesh->add_elem(new Quad4);
+      elem_Quad4->set_node(0) = nodes[node_id_array[i][j]];
+      elem_Quad4->set_node(1) = nodes[node_id_array[i + 1][j]];
+      elem_Quad4->set_node(2) = nodes[node_id_array[i + 1][(j + 1) % input_ext_node_num]];
+      elem_Quad4->set_node(3) = nodes[node_id_array[i][(j + 1) % input_ext_node_num]];
+      elem_Quad4->subdomain_id() = _peripheral_ring_block_id;
+
+      if (i == 0)
+        boundary_info.add_side(elem_Quad4, 3, OUTER_SIDESET_ID_ALT);
+      if (i == _peripheral_layer_num - 1)
+        boundary_info.add_side(elem_Quad4, 1, OUTER_SIDESET_ID);
+    }
+
+  // This would make sure that the boundary OUTER_SIDESET_ID is deleted after stitching.
+  if (_input_mesh_external_bid != OUTER_SIDESET_ID)
+    MooseMesh::changeBoundaryId(*input_mesh, _input_mesh_external_bid, OUTER_SIDESET_ID, true);
+  mesh->prepare_for_use();
+  mesh->stitch_meshes(*input_mesh, OUTER_SIDESET_ID_ALT, _input_mesh_external_bid, TOLERANCE, true);
+
+  // Assign subdomain name to the new block if applicable
+  if (isParamValid("peripheral_ring_block_name"))
+    mesh->subdomain_name(_peripheral_ring_block_id) = _peripheral_ring_block_name;
+  // Assign customized external boundary id
+  if (_external_boundary_id > 0)
+    MooseMesh::changeBoundaryId(*mesh, OUTER_SIDESET_ID, _external_boundary_id, false);
+  // Assign customized external boundary name
+  if (!_external_boundary_name.empty())
+  {
+    mesh->boundary_info->sideset_name(
+        _external_boundary_id > 0 ? _external_boundary_id : (boundary_id_type)OUTER_SIDESET_ID) =
+        _external_boundary_name;
+    mesh->boundary_info->nodeset_name(
+        _external_boundary_id > 0 ? _external_boundary_id : (boundary_id_type)OUTER_SIDESET_ID) =
+        _external_boundary_name;
+  }
+
+  return dynamic_pointer_cast<MeshBase>(mesh);
+}
+
+bool
+PeripheralRingMeshGenerator::isBoundaryValid(ReplicatedMesh & mesh,
+                                             Real & max_node_radius,
+                                             unsigned short & invalid_type,
+                                             const Point origin_pt,
+                                             const boundary_id_type bid) const
+{
+  max_node_radius = 0.0;
+  invalid_type = 0;
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list_tmp = boundary_info.build_side_list();
+  unsigned int elem_counter = 0;
+  std::vector<std::pair<dof_id_type, dof_id_type>> boundary_node_assm;
+  std::vector<dof_id_type> boundary_orderred_node_list;
+  bool isFlipped = false;
+  for (unsigned int i = 0; i < side_list_tmp.size(); i++)
+  {
+    if (std::get<2>(side_list_tmp[i]) == bid)
+    {
+      elem_counter++;
+      // store two nodes of each side

```

Comment:
Do we have tests + moose errors anywhere here for the `elem->dim() != 2` case?  It's obvious from the docs that only 2D meshes should be supported, but if someone messes up and tries to use this with a 3D or mixed-dimension mesh anyway we'd like them to get a sane error message out.